tests/mir_test_framework/window_management_test_harness.cpp

```diff
@@ -130,12 +297,72 @@ class mir_test_framework::WindowManagementTestHarness::Self final : public ms::S
         pending_actions.clear();
     }
 
-    mir::Server& server;
+    class SurfaceObserver final : public ms::SurfaceObserver
+    {
+    public:
+        explicit SurfaceObserver(Self* self)
+            : self(self) {}
```

Comment:
It would separate concerns better to pass in a `std::function<void(ms::Surface const* surf)>` that invokes the logic on `Self`?

Then the construction site becomes something like:

```c++
    explicit Self(mir::Server& server)
        : server(server),
          surface_observer(std::make_shared<SurfaceObserver>([this])(ms::Surface const* surf) { client_surface_close_requested(surf); })
```

And the `client_surface_close_requested()` logic is implemented in `Self`